src/env-inl.h

```diff
@@ -979,6 +979,14 @@ void Environment::ForEachBaseObject(T&& iterator) {
   }
 }
 
+bool AsyncRequest::is_stopped() const {
+  return stopped_.load(std::memory_order_relaxed);
+}
+
+void AsyncRequest::set_stopped(bool flag) {
+  stopped_.store(flag, std::memory_order_relaxed);
```

Comment:
so: the heavy usage of muexes around the worker code where multi-thread data access was expected, was almost always to ensure data consistency by flushing cache lines? (in other words, writes in one thread is made visible to other threads instantly). If so, `std::memory_order_relaxed` constraint is insufficient to ensure that? we might need at least `memory_order_acquire` ?

Reply:
I guess I retract above question - when I tested with a small code, I see `mfence` or `sync` instruction being added with `memory_order_relaxed` itself; so pls ignore.

Reply:
So … my line of thinking was that the syscalls behind `uv_async_send()` would themselves present full memory barriers. I’ll check later and verify that that’s indeed correct.

Reply:
@addaleax - that may be true; but we do have `set_stopped` `is_stopped` calls that the threads can call but do not involve syscalls? however:

```c++
int foo(bool flag) {
  stopped_.store(flag, std::memory_order_relaxed);
}
```
this is what I tested and this is what I see in the generated code:

```gdb
(gdb) x/30i 0x40053e
   0x40053e <_ZNSt11atomic_bool5storeEbSt12memory_order>:	push   rbp
   0x40053f <_ZNSt11atomic_bool5storeEbSt12memory_order+1>:	mov    rbp,rsp
   0x400542 <_ZNSt11atomic_bool5storeEbSt12memory_order+4>:	sub    rsp,0x30
   0x400546 <_ZNSt11atomic_bool5storeEbSt12memory_order+8>:	mov    QWORD PTR [rbp-0x28],rdi
   0x40054a <_ZNSt11atomic_bool5storeEbSt12memory_order+12>:	mov    eax,esi
   0x40054c <_ZNSt11atomic_bool5storeEbSt12memory_order+14>:	mov    DWORD PTR [rbp-0x30],edx
   0x40054f <_ZNSt11atomic_bool5storeEbSt12memory_order+17>:	mov    BYTE PTR [rbp-0x2c],al
   0x400552 <_ZNSt11atomic_bool5storeEbSt12memory_order+20>:	movzx  eax,BYTE PTR [rbp-0x2c]
   0x400556 <_ZNSt11atomic_bool5storeEbSt12memory_order+24>:	mov    rdx,QWORD PTR [rbp-0x28]
   0x40055a <_ZNSt11atomic_bool5storeEbSt12memory_order+28>:	mov    QWORD PTR [rbp-0x8],rdx
   0x40055e <_ZNSt11atomic_bool5storeEbSt12memory_order+32>:	mov    BYTE PTR [rbp-0x9],al
   0x400561 <_ZNSt11atomic_bool5storeEbSt12memory_order+35>:	and    BYTE PTR [rbp-0x9],0x1
   0x400565 <_ZNSt11atomic_bool5storeEbSt12memory_order+39>:	mov    eax,DWORD PTR [rbp-0x30]
   0x400568 <_ZNSt11atomic_bool5storeEbSt12memory_order+42>:	mov    DWORD PTR [rbp-0x10],eax
   0x40056b <_ZNSt11atomic_bool5storeEbSt12memory_order+45>:	mov    eax,DWORD PTR [rbp-0x10]
   0x40056e <_ZNSt11atomic_bool5storeEbSt12memory_order+48>:	mov    esi,0xffff
   0x400573 <_ZNSt11atomic_bool5storeEbSt12memory_order+53>:	mov    edi,eax
   0x400575 <_ZNSt11atomic_bool5storeEbSt12memory_order+55>:	
    call   0x40052a <_ZStanSt12memory_orderSt23__memory_order_modifier>
   0x40057a <_ZNSt11atomic_bool5storeEbSt12memory_order+60>:	mov    DWORD PTR [rbp-0x14],eax
   0x40057d <_ZNSt11atomic_bool5storeEbSt12memory_order+63>:	movzx  edx,BYTE PTR [rbp-0x9]
   0x400581 <_ZNSt11atomic_bool5storeEbSt12memory_order+67>:	mov    rax,QWORD PTR [rbp-0x8]
   0x400585 <_ZNSt11atomic_bool5storeEbSt12memory_order+71>:	mov    BYTE PTR [rax],dl
   0x400587 <_ZNSt11atomic_bool5storeEbSt12memory_order+73>:	mfence 
   0x40058a <_ZNSt11atomic_bool5storeEbSt12memory_order+76>:	leave  
   0x40058b <_ZNSt11atomic_bool5storeEbSt12memory_order+77>:	ret    
```

please note that `mfence` at `0x400587` that settles the matter?

Reply:
> please note that `mfence` at `0x400587` that settles the matter?

On x64 it does, yes – to be honest, I don’t know how the different memory order modes are implemented on different platforms? It seems like this disassembled implementation simply ignores the order argument?

Reply:
```gdb
   0x400529 <_Z3foob+35>:	ret    
   0x40052a <_ZStanSt12memory_orderSt23__memory_order_modifier>:	push   rbp
```

@addaleax - if you look at the continuity of the instructions,  looks like these (the atomic* helpers) are not static APIs, but compiler-generated code, on the fly; so it is possible that only necessary code was generated, on a per compilation unit basis?

Reply:
> my line of thinking was that the syscalls behind uv_async_send() would themselves present full memory barriers.

Libuv doesn't promise that. `uv_async_send()` can (at least in theory) elide the system call.

I'm kind of surprised the compiler emits an mfence. It's not needed on x64 (nor any other architecture, I think?) because aligned loads and stores are always atomic. It might just be a compiler bug; I wouldn't depend on it.

Reply:
@gireeshpunathil I wouldn’t think so, the linker should be able to elide multiple copies of that variable into a single one.

@bnoordhuis Yeah, thanks. I’ve removed the `memory_order_relaxed` bit.